<a href="https://colab.research.google.com/github/peremartra/Large-Language-Model-Notebooks-Course/blob/main/P2-MHF/Aligning_DPO_open_llama_v2.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

<div align="center">
<h1><a href="https://github.com/peremartra/Large-Language-Model-Notebooks-Course">Learn by Doing LLM Projects</a></h1>
    <h3>Understand And Apply Large Language Models</h3>
    <h2>Creating and Publishing Your Own LLM.</h2>
    <h3>Aligning with DPO a phi3-3 model.</h3>
    by <b>Pere Martra</b>
</div>

<br>

<div align="center">
    &nbsp;
    <a target="_blank" href="https://www.linkedin.com/in/pere-martra/"><img src="https://img.shields.io/badge/style--5eba00.svg?label=LinkedIn&logo=linkedin&style=social"></a>
    
</div>

<br>
<hr>


In this Notebook we are going to align a Microsoft Phi-3 Model using DPO, and publish it to Hugging Face!

Base Model: https://huggingface.co/microsoft/Phi-3-mini-4k-instruct

Dataset Used: https://huggingface.co/datasets/argilla/distilabel-capybara-dpo-7k-binarized

Model obtained: https://huggingface.co/oopere/martra-phi-3-mini-dpo

## To start, a brief introduction to DPO.

The revolution we're currently experiencing around Large Language Models began with the emergence of ChatGPT and its GPT-3.5 model.

Something different had been done with GPT-3.5, which was actually a derivative of GPT-3, a model that did not generate nearly as much excitement as its successor.

Many people, including myself, believe that the main difference was the use of Alignemet using RLHF - Reinforcement Learning from Human Feedback.

Nowadays RLHF has been displaced by a technique that achieves the same result in a much more efficient way: DPO - Direct Preference Optimization.

Both DPO and RLHF are alignment techniques that require a dataset containing correct and incorrect responses to the same prompt.

But from here, the differences begin. RLHF uses this dataset to train a second model, called a reward model, which will be used in the alignment process. DPO, on the other hand, uses the dataset directly to train the final model. This is the main difference between the two techniques.

As you can imagine, DPO is a more direct technique that requires fewer resources. When we're talking about models with tens of billions of parameters, any reduction in resource consumption can result in significant cost savings.

The implementation of DPO that we are going to use is the one developed by Hugging Face in their TRL library, which stands for Transformer Reinforcement Learning. DPO can be considered a reinforcement learning technique, where the model is rewarded during its training phase based on its responses.

_______________________

Since is necesary to save the created model, the notebook mounts a disk in  Google Drive. If you run it locally, you don't need to execute this line of code. You can actually also run it on Google Colab without mounting a disk in your drive, but then every time you close the session you'll lose the saved model, as it will be saved in the temporary directory of the Google Colab session.

In [2]:
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


Now that the disk is mounted, it's time to load the necessary libraries:

The only one that might be new to you is the trl library, which stands for Transformer Reinforcement Learning. You'll be importing the DPOTrainer class from this library, which you'll use to perform the DPO fine-tuning of the model.

In [3]:
!pip install -q datasets==2.19.1
!pip install -q trl==0.8.6
!pip install -q peft==0.11.1
!pip install -q transformers==4.41.0
!pip install -q bitsandbytes==0.43.1
!pip install -q sentencepiece==0.1.99
!pip install -q accelerate==0.30.1
!pip install -q huggingface_hub==0.23.1

     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 542.0/542.0 kB 3.1 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 116.3/116.3 kB 16.8 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 194.1/194.1 kB 19.7 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 134.8/134.8 kB 17.0 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 245.2/245.2 kB 1.8 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 314.1/314.1 kB 7.0 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 103.4/103.4 kB 15.4 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 21.3/21.3 MB 52.4 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 251.6/251.6 kB 1.9 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 9.1/9.1 MB 25.3 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 119.8/119.8 MB 10.0 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 302.6/302.6 kB 2.3 MB/s eta 0:00:00
     ━━━━━━━━━━

In [4]:
#Import necessary classes.
import gc
import torch

import transformers
from transformers import AutoModelForCausalLM, AutoTokenizer
from transformers import TrainingArguments, BitsAndBytesConfig
from datasets import load_dataset
from peft import LoraConfig, get_peft_model, PeftModel
from trl import DPOTrainer
import bitsandbytes as bnb

from getpass import getpass


Another necessary step is to log in to Hugging Face.

In [5]:
hf_token = getpass("Hugging Face: ")

Hugging Face: ··········


In [6]:
!huggingface-cli login --token $hf_token

The token has not been saved to the git credentials helper. Pass `add_to_git_credential=True` in this function directly or `--add-to-git-credential` if using via `huggingface-cli` if you want to set the git credential as well.
Token is valid (permission: write).
Your token has been saved to /root/.cache/huggingface/token
Login successful


## Format dataset

The model I've chosen is the Microsoft Phi-3 mini with the 4k context. It's a 3.8B parameter model that is very competitive and in many cases outperforms 7B parameter models.
I've chosen a small model so that its training can be done with few resources on Google Colab or on a not very large GPU.


In [7]:
model_name = "openlm-research/open_llama_3b_v2"
new_model = "martra-open-llamav2-dpo-3b"

In [8]:
# Tokenizer
tokenizer = AutoTokenizer.from_pretrained(model_name,
                                          use_fast=False)
tokenizer.pad_token = tokenizer.eos_token

tokenizer_config.json:   0%|          | 0.00/593 [00:00<?, ?B/s]

tokenizer.model:   0%|          | 0.00/512k [00:00<?, ?B/s]

special_tokens_map.json:   0%|          | 0.00/330 [00:00<?, ?B/s]

You are using the default legacy behaviour of the <class 'transformers.models.llama.tokenization_llama.LlamaTokenizer'>. This is expected, and simply means that the `legacy` (previous) behavior will be used so nothing changes for you. If you want to use the new behaviour, set `legacy=False`. This should only be set if you understand what it means, and thoroughly read the reason why this was added as explained in https://github.com/huggingface/transformers/pull/24565


Before you begin training the model, is necesary need to load the dataset and transform it to fit the format required by the DPOTrainer class, that consists of three fields: the prompt, the chosen answer, and a discarded answer.

I'm loading just a fer rows of the Dataset, feel free to use the full Dataset if you have enough time.

You may need more or less 2 hours of a A100 GPU to train with Full Dataset for a 20 epochs. If you use the configuration in the notebook you are going to need just 40 minutes in a L4 GPU.

In [9]:
# Load dataset
dataset_original =  load_dataset("argilla/distilabel-capybara-dpo-7k-binarized",
                                 split='train[300:2500]')
dataset_eval = load_dataset("argilla/distilabel-capybara-dpo-7k-binarized",
                            split='train[:300]')

# Save columns
original_columns = dataset_original.column_names
print(original_columns)

Generating train split:   0%|          | 0/7563 [00:00<?, ? examples/s]

['source', 'conversation', 'original_response', 'generation_prompt', 'raw_generation_responses', 'new_generations', 'prompt', 'chosen', 'rejected', 'rating_chosen', 'rating_rejected', 'chosen_model', 'rejected_model']


In [10]:
dataset_original

Dataset({
    features: ['source', 'conversation', 'original_response', 'generation_prompt', 'raw_generation_responses', 'new_generations', 'prompt', 'chosen', 'rejected', 'rating_chosen', 'rating_rejected', 'chosen_model', 'rejected_model'],
    num_rows: 2200
})

The dataset has many more columns than are actually necessary for the DPO process. However, I'm going to take advantage of a couple of them to filter the data to be used.

In [11]:
dataset_filtered = dataset_original.filter(
  lambda r: r["rating_chosen"]>=4.5 and r["rating_rejected"] <= 2.5
)

Filter:   0%|          | 0/2200 [00:00<?, ? examples/s]

This first filter only retrieves those rows where the rating of the chosen response is very high and the rating of the discarded responses is very low. This is a way to facilitate the model's learning, although it's also possible that it doesn't help in the last epochs of training.

I'm going to perform a second filter to keep the prompt length under control, as the selected model only accepts a length of 4000 tokens.

In [12]:
dataset_filtered = dataset_filtered.map(lambda r: {"messages": len(r["chosen"])}).filter(lambda r: r["messages"]<3 and len(r["prompt"]) + len(r["chosen"]) + len(r["rejected"]) < 3800)


Map:   0%|          | 0/400 [00:00<?, ? examples/s]

Filter:   0%|          | 0/400 [00:00<?, ? examples/s]

In [13]:
dataset_filtered

Dataset({
    features: ['source', 'conversation', 'original_response', 'generation_prompt', 'raw_generation_responses', 'new_generations', 'prompt', 'chosen', 'rejected', 'rating_chosen', 'rating_rejected', 'chosen_model', 'rejected_model', 'messages'],
    num_rows: 169
})

The dataset still have all the columns, but the number of rows has been significantly reduced. Let me warn you that 169 rows are too few to perform proper training; again, this reduction is so that will be possible to execute the notebook in just some minutes, and obtain results.

In [14]:
#Repeat the same filters with the Validation Dataset.
dataset_eval_filtered = dataset_eval.filter(
  lambda r: r["rating_chosen"]>=4.5 and r["rating_rejected"] <= 2.5
)
dataset_eval_filtered = dataset_eval_filtered.map(lambda r: {"messages": len(r["chosen"])}).filter(lambda r: r["messages"]<3 and len(r["prompt"]) + len(r["chosen"]) + len(r["rejected"]) < 3800)
dataset_eval_filtered

Filter:   0%|          | 0/300 [00:00<?, ? examples/s]

Map:   0%|          | 0/51 [00:00<?, ? examples/s]

Filter:   0%|          | 0/51 [00:00<?, ? examples/s]

Dataset({
    features: ['source', 'conversation', 'original_response', 'generation_prompt', 'raw_generation_responses', 'new_generations', 'prompt', 'chosen', 'rejected', 'rating_chosen', 'rating_rejected', 'chosen_model', 'rejected_model', 'messages'],
    num_rows: 28
})

Now, it's a matter of creating a function to adapt the dataset's structure to what's required by the DPOTraining class.

I have to confess that I've cheated a little bit. The function comes from the Hugging Face dataset card. I only had to remove an error that they had missed.

In summary, the function takes a row and retrieves only the three necessary columns. It also applies a small format to the responses, which I've adapted to the format required by the Model, adding the label <|end|> after the responses.


In [15]:
def chatml_format(example):
    # get everything except the last message as input
    prompt = tokenizer.apply_chat_template(example["chosen"][:-1], tokenize=False,
                                           add_generation_prompt=True)
    # get the last assistant responses
    chosen = example["chosen"][-1]["content"] + "<|end|>\n"
    rejected = example["rejected"][-1]["content"] + "<|end|>\n"

    return {
        "prompt": prompt,
        "chosen": chosen,
        "rejected": rejected,
    }

In [16]:
original_columns

['source',
 'conversation',
 'original_response',
 'generation_prompt',
 'raw_generation_responses',
 'new_generations',
 'prompt',
 'chosen',
 'rejected',
 'rating_chosen',
 'rating_rejected',
 'chosen_model',
 'rejected_model']

I'll use the dataset's map function to execute the transformation on each row, and also remove the original columns.

In [17]:
# Format dataset
dataset = dataset_filtered.map(
    chatml_format,
    remove_columns=dataset_filtered.column_names
)
# Print sample
dataset[12]

Map:   0%|          | 0/169 [00:00<?, ? examples/s]

No chat template is set for this tokenizer, falling back to a default class-level template. This is very error-prone, because models are often trained with templates different from the class default! Default chat templates are a legacy feature and will be removed in Transformers v4.43, at which point any code depending on them will stop working. We recommend setting a valid chat template before then to ensure that this model continues working without issues.


{'prompt': '<s>[INST] Solve 36146684-304553543134. Only respond with math and no words. [/INST]',
 'chosen': '36146684 - 304553543134 = -304517396450<|end|>\n',
 'rejected': '(36146684 / 3134) * (36146684 mod 3134) + (30455354 / 17) * (30455354 mod 17) = 11415845286790903\nAlternatively, using prime factorization and the Chinese Remainder Theorem:\n36146684 = 2^5 * 9573, 30455354 = 2 * 29 * 4171\n36146684 mod 9573 = 4332, 30455354 mod 29 = 13, 30455354 mod 4171 = 3965\n(36146684 / 9573) * 4332 + (30455354 / 29) * 13 + (30455354 / 4171) * 3965 = 11415845286790903<|end|>\n'}

In [18]:
# Format dataset
dataset_eval = dataset_eval_filtered.map(
    chatml_format,
    remove_columns=original_columns
)

Map:   0%|          | 0/28 [00:00<?, ? examples/s]

In [19]:
# Print sample
dataset_eval[2]

{'prompt': '<s>[INST] Assist me in calculating 9319357631 plus 595. Numbers and symbols only, please. [/INST]',
 'chosen': 'The sum of 9319357631 and 595 is 9319358226.<|end|>\n',
 'rejected': 'The result of adding 9319357631 and 595 is 9319363626.<|end|>\n',
 'messages': 2}

The format is adapted to chat in phi3:
https://huggingface.co/microsoft/Phi-3-mini-128k-instruct



<|user|>
I am going to Paris, what should I see?<|end|>

<|assistant|>
Paris, the capital of France, is known for its stunning architecture, art museums, historical landmarks, and romantic atmosphere. Here are some of the top attractions to see in Paris:\n\n1. The Eiffel Tower: The iconic Eiffel Tower is one of the most recognizable landmarks in the world and offers breathtaking views of the city.\n2. The Louvre Museum: The Louvre is one of the world's largest and most famous museums, housing an impressive collection of art and artifacts, including the Mona Lisa.\n3. Notre-Dame Cathedral: This beautiful cathedral is one of the most famous landmarks in Paris and is known for its Gothic architecture and stunning stained glass windows.\n\nThese are just a few of the many attractions that Paris has to offer. With so much to see and do, it's no wonder that Paris is one of the most popular tourist destinations in the world."<|end|>

<|user|>
What is so great about #1?<|end|>



In [20]:
dataset

Dataset({
    features: ['prompt', 'chosen', 'rejected'],
    num_rows: 169
})

In [21]:
dataset_eval

Dataset({
    features: ['prompt', 'chosen', 'rejected', 'messages'],
    num_rows: 28
})

## Train model with DPO

## Finetuning with DPOTrainer.

Time to start working with the necessary configurations to perform alignment using DPO.



In [22]:
# LoRA configuration
peft_config = LoraConfig(
    r=16,
    lora_alpha=32,
    lora_dropout=0.05,
    bias="none",
    task_type="CAUSAL_LM",
    #target_modules=['o_proj', 'qkv_proj'] #phi-3
    target_modules="all-linear"
)

The value of **r** indicates the size of the reparameterization; the higher the value, the more parameters are trained. An 8 is at the upper limit of what is recommended for small models.

To further accentuate the weight of the new training, I use the **lora_alpha** value. It's a multiplier that adjusts the layers inserted by LoRA. Normally it's left at 1, but in the case of DPO, I've seen values as high as 128.

The recommendation is that **lora_alpha** should be double the value of **r**. Since **r** varies depending on the model size, you may end up with a very high lora_alpha value if you want to fine-tune a large model and, for example, specify an **r** of 64.

In [23]:
bnb_config = BitsAndBytesConfig(
    load_in_4bit=True,
    bnb_4bit_use_double_quant=True,
    bnb_4bit_quant_type="nf4",
    bnb_4bit_compute_dtype=torch.float16
)

The quantization configuration holds no secrets, we are reducing the model's precision to 4 bits.

In [24]:
# Model to fine-tune
model = AutoModelForCausalLM.from_pretrained(
    model_name,
    quantization_config=bnb_config
)
model.config.use_cache = False

config.json:   0%|          | 0.00/506 [00:00<?, ?B/s]

`low_cpu_mem_usage` was None, now set to True since model is quantized.


pytorch_model.bin:   0%|          | 0.00/6.85G [00:00<?, ?B/s]

generation_config.json:   0%|          | 0.00/137 [00:00<?, ?B/s]

The next step is to create the training parameters.

In [25]:
# Training arguments
training_args = TrainingArguments(
    per_device_train_batch_size=2,
    per_device_eval_batch_size=2,
    gradient_accumulation_steps=2,
    gradient_checkpointing=True,
    remove_unused_columns=True,
    learning_rate=5.0e-06,
    evaluation_strategy="epoch",
    logging_strategy="epoch",
    lr_scheduler_type="cosine",
    num_train_epochs=6,
    save_strategy="epoch",
    logging_steps=1,
    output_dir=new_model,
    optim="paged_adamw_32bit",
    warmup_steps=2,
    bf16=True,
    report_to="none",
)


/usr/local/lib/python3.10/dist-packages/transformers/training_args.py:1474: FutureWarning: `evaluation_strategy` is deprecated and will be removed in version 4.46 of 🤗 Transformers. Use `eval_strategy` instead
  warnings.warn(


I'll try to explain the value of the most important and specific ones.

**lr_scheduler_type**="cosine": The learning rate is adjusted according to a cosine schedule. It starts at the value specified in learning_rate and then gradually decreases.

**warmup_steps**=2:  For the first two epochs, the learning rate is adjusted by increasing its value instead of decreasing it. The aim is to stabilize the learning process.

**Gradient_accumulation_steps**=2: To save memory. I accumulate the gradients over two steps before updating the model weights.

With these parameters, I've tried to find a training setup with low memory requirements, thanks to the use of gradient accumulation, gradient checkpointing, a small batch size, and the use of bf16 along with the paged_adamw_32bit optimizer.

In [26]:
# Create DPO trainer
trainer = DPOTrainer(
    model,
    args=training_args,
    train_dataset=dataset,
    eval_dataset=dataset_eval,
    tokenizer=tokenizer,
    peft_config=peft_config,
    beta=0.1,
    max_prompt_length=2048,
    max_length=2048,
)

/usr/local/lib/python3.10/dist-packages/trl/trainer/dpo_trainer.py:332: UserWarning: When using DPODataCollatorWithPadding, you should set `remove_unused_columns=False` in your TrainingArguments we have set it for you, but you should do it yourself in the future.
  warnings.warn(


Map:   0%|          | 0/169 [00:00<?, ? examples/s]

Map:   0%|          | 0/28 [00:00<?, ? examples/s]

The parameters for DPOTrainer are quite simple. You need to pass it the configurations you've created, the two evaluation datasets, and the maximum length of the prompt and response.

The indicated beta value is a standard that balances the new training with the model's base knowledge. If you want the new training to have more weight, perhaps because you're training for a very specific task, you could specify a lower beta value.

In [27]:
# Fine-tune model with DPO
trainer.train()

/usr/local/lib/python3.10/dist-packages/torch/utils/checkpoint.py:464: UserWarning: torch.utils.checkpoint: the use_reentrant parameter should be passed explicitly. In version 2.4 we will raise an exception if use_reentrant is not passed. use_reentrant=False is recommended, but if you need to preserve the current default behavior, you can pass use_reentrant=True. Refer to docs for more details on the differences between the two variants.
  warnings.warn(
/usr/local/lib/python3.10/dist-packages/torch/utils/checkpoint.py:91: UserWarning: None of the inputs have requires_grad=True. Gradients will be None
  warnings.warn(
Could not estimate the number of tokens of the input, floating-point operations will not be computed


/usr/local/lib/python3.10/dist-packages/huggingface_hub/file_download.py:1132: FutureWarning: `resume_download` is deprecated and will be removed in version 1.0.0. Downloads always resume when possible. If you want to force a new download, use `force_download=True`.
  warnings.warn(
/usr/local/lib/python3.10/dist-packages/torch/utils/checkpoint.py:464: UserWarning: torch.utils.checkpoint: the use_reentrant parameter should be passed explicitly. In version 2.4 we will raise an exception if use_reentrant is not passed. use_reentrant=False is recommended, but if you need to preserve the current default behavior, you can pass use_reentrant=True. Refer to docs for more details on the differences between the two variants.
  warnings.warn(
/usr/local/lib/python3.10/dist-packages/torch/utils/checkpoint.py:91: UserWarning: None of the inputs have requires_grad=True. Gradients will be None
  warnings.warn(
/usr/local/lib/python3.10/dist-packages/huggingface_hub/file_download.py:1132: FutureWarni

TrainOutput(global_step=252, training_loss=0.4524516567351326, metrics={'train_runtime': 571.5925, 'train_samples_per_second': 1.774, 'train_steps_per_second': 0.441, 'total_flos': 0.0, 'train_loss': 0.4524516567351326, 'epoch': 5.929411764705883})

It seems to have worked reasonably well, although there might be a potential overfitting issue, where the model adapts better to the training data than to the evaluation data. To mitigate overfitting, you could expand the dataset and try increasing the **lora_dropout** parameter in **LoraConfig**.


## Upload model

In [28]:
#PATH_MODEL="/content/drive/MyDrive/final_checkpoint"
PATH_MODEL="/content/drive/MyDrive/apress_checkpoint"

In [29]:
# Save artifacts
trainer.model.save_pretrained(PATH_MODEL)
tokenizer.save_pretrained(PATH_MODEL)



/usr/local/lib/python3.10/dist-packages/huggingface_hub/file_download.py:1132: FutureWarning: `resume_download` is deprecated and will be removed in version 1.0.0. Downloads always resume when possible. If you want to force a new download, use `force_download=True`.
  warnings.warn(


('/content/drive/MyDrive/apress_checkpoint/tokenizer_config.json',
 '/content/drive/MyDrive/apress_checkpoint/special_tokens_map.json',
 '/content/drive/MyDrive/apress_checkpoint/tokenizer.model',
 '/content/drive/MyDrive/apress_checkpoint/added_tokens.json')

Execute this cell only if you are having memory issues. (Not you, of course, I mean your environment 🤗).

In [30]:
#Flush memory
#del dpo_trainer, model
gc.collect()
torch.cuda.empty_cache()

Now, you're going to load the original model again, but this time in its unquantized format.

In [32]:
base_model = AutoModelForCausalLM.from_pretrained(
    model_name,
    return_dict=True,
    torch_dtype=torch.bfloat16,
)
tokenizer = AutoTokenizer.from_pretrained(model_name,
                                          use_fast=False)

The original model and the saved training are being merged.

In [33]:
model = PeftModel.from_pretrained(base_model, PATH_MODEL)
model = model.merge_and_unload()

 The model that you have in memory is now a combination of the base model and the adapter that you have trained. You can now save this new model and upload it to Hugging Face.

In [34]:
model.save_pretrained(new_model)
tokenizer.save_pretrained(new_model)

('martra-open-llamav2-dpo-3b/tokenizer_config.json',
 'martra-open-llamav2-dpo-3b/special_tokens_map.json',
 'martra-open-llamav2-dpo-3b/tokenizer.model',
 'martra-open-llamav2-dpo-3b/added_tokens.json')

In [35]:
model.push_to_hub(new_model,
                  private=True,
                  use_temp_dir=False,
                  token=hf_token)
tokenizer.push_to_hub(new_model,
                      private=True,
                      use_temp_dir=False,
                      token=hf_token)

model-00002-of-00002.safetensors:   0%|          | 0.00/1.86G [00:00<?, ?B/s]

model-00001-of-00002.safetensors:   0%|          | 0.00/5.00G [00:00<?, ?B/s]

Upload 2 LFS files:   0%|          | 0/2 [00:00<?, ?it/s]

README.md:   0%|          | 0.00/5.17k [00:00<?, ?B/s]

tokenizer.model:   0%|          | 0.00/512k [00:00<?, ?B/s]

CommitInfo(commit_url='https://huggingface.co/oopere/martra-open-llamav2-dpo-3b/commit/699d77717b381c11357830fce35f04f3af84cced', commit_message='Upload tokenizer', commit_description='', oid='699d77717b381c11357830fce35f04f3af84cced', pr_url=None, pr_revision=None, pr_num=None)

## Inference

Let's test the new model and compare with the original

In [36]:
# Format prompt
message = [
    {"role": "user", "content": "3713841893836/4?\nLimit your response to mathematical expressions and symbols."}
]

In [37]:
#new_model="oopere/martra-test-phi-3-mini-dpo"
tokenizer_new_model = AutoTokenizer.from_pretrained(new_model)
prompt = tokenizer_new_model.apply_chat_template(message, add_generation_prompt=True, tokenize=False)

# Create pipeline
pipeline_new = transformers.pipeline(
    "text-generation",
    model=new_model,
    tokenizer=tokenizer_new_model
)

You set `add_prefix_space`. The tokenizer needs to be converted from the slow tokenizers


Loading checkpoint shards:   0%|          | 0/2 [00:00<?, ?it/s]

In [38]:
# Generate text
sequences = pipeline_new(
    prompt,
    do_sample=True,
    temperature=0.1,
    top_p=0.2,
    num_return_sequences=1,
    max_length=200,
)
print(sequences[0]['generated_text'])

Truncation was not explicitly activated but `max_length` is provided a specific value, please use `truncation=True` to explicitly truncate examples to max length. Defaulting to 'longest_first' truncation strategy. If you encode pairs of sequences (GLUE-style) with the tokenizer you can select this strategy more precisely by providing a specific strategy to `truncation`.


<s>[INST] 3713841893836/4?
Limit your response to mathematical expressions and symbols. [/INST]
[INST] 3713841893836/4?[/INST]
[INST] 3713841893836/4?[/INST]
[INST] 3713841893836/4?[/INST]
[INST] 3713841893836/4?[/INST]
[INST] 3713841893836/4?[/INST]
[INST] 3713841893836/4?[/INST]
[INST] 3713841893836/4?


PERFECT! The new model only returns numbers, as we want!

If you don't want to wait to the training, just test my model on hugging Face. It has been trained with the same Dataset for 2 hours in a A100 GPU on Colab.

In [ ]:
#Test my Model on hugging Face.
new_model="oopere/martra-phi-3-mini-dpo"
tokenizer_new_model = AutoTokenizer.from_pretrained(new_model)
prompt = tokenizer_new_model.apply_chat_template(message, add_generation_prompt=True, tokenize=False)

# Create pipeline
pipeline_new = transformers.pipeline(
    "text-generation",
    model=new_model,
    tokenizer=tokenizer_new_model
)



Special tokens have been added in the vocabulary, make sure the associated word embeddings are fine-tuned or trained.


Loading checkpoint shards:   0%|          | 0/2 [00:00<?, ?it/s]

In [ ]:
# Generate text
sequences = pipeline_new(
    prompt,
    do_sample=True,
    temperature=0.1,
    top_p=0.2,
    num_return_sequences=1,
    max_length=200,
)
print(sequences[0]['generated_text'])

Truncation was not explicitly activated but `max_length` is provided a specific value, please use `truncation=True` to explicitly truncate examples to max length. Defaulting to 'longest_first' truncation strategy. If you encode pairs of sequences (GLUE-style) with the tokenizer you can select this strategy more precisely by providing a specific strategy to `truncation`.


<s><|user|>
3713841893836/4?
Limit your response to mathematical expressions and symbols.<|end|>
<|assistant|>
 3713841893836 / 4 = 928460473459


# Summary

The model alignment process has been a complete success. The truth is, with the Hugging Face libraries, everything is straightforward.

The challenge is knowing about the technique, when to apply it, and having the necessary data.

In this notebook, you've addressed the first two points.

I got a lot of inspiration from:

* [RLHF in 2024 with DPO & Hugging Face](https://www.philschmid.de/dpo-align-llms-in-2024-with-trl) by Phil Schmid.

* [Fine-tune a Mistral-7b model with Direct Preference Optimizatio](https://medium.com/towards-data-science/fine-tune-a-mistral-7b-model-with-direct-preference-optimization-708042745aac) by Maxime Labonne

